### ISSUE TITLE
More advanced local server?

### ISSUE THREAD

**Issue by astronouth7303** [r0x0r/pywebview#483]
### Specification
- pywebview version: 3.2
- platform / version: Linux+gtk

### Description  
  
The built-in local server is fine for really basic stuff, but:

* You can't use `importlib.resources` with it
* You can't "mount" several directories
* You can't fallback to something else

Since `wsgiref` has been part of the standard library for a while, doing some kind of basic wsgi routing middleware is doable without additional dependencies.

The basic concept is that you "mount" various sources at directory prefixes, and it'll select the most appropriate one. Sources include:

* A filesystem directory
* A python package (data loaded via `importlib.resources`)
* Another WSGI app

I think these three represent the most common uses while providing progressive enhancement. (It's very easy to "scale up" with this system as an application becomes more complex, without hitting any walls.)

The routing algorithm is simple: Use the longest matching directory prefix. "Directory prefix" means that `/foo/bar` will match `/foo/bar`, `/foo/bar/` and `/foo/bar/spam` but not `/foo/bar-spam`. Longest means that you can mount `/` and `/static` and it'll work as you expect.

I think package data (`importlib.resources`) should be the preferred way of loading application assets (although I haven't seen how it works with py2exe and friends). Directories are useful for serving user data, generated files, cache directories, etc. WSGI apps just allow open-ended anything.
  
### Practicalities

- YES I am willing to work on this issue myself.

- NO I am prepared to support this issue financially.

**Comment by astronouth7303** [comment]
(Under the hood, both directory and package static serving could also just be WSGI apps.)

**Comment by r0x0r** [comment]
Yes, I agree Python's own web server is not very good and I have been
pondering about replacing it with something else. I have had something
like bottle in mind, but introducing a foreign dependency is always
risky.
I am not familiar with wsgiref, but it sounds promising. Do you know
if it is able to server media files? Python's web server, for one, is
not able to to serve videos on GTK.

**Comment by astronouth7303** [comment]
I'd have to check. I'd probably do the actual file serving code based on [whitenoise](http://whitenoise.evans.io/) (not as a dep, though, i believe there's some limitations it makes for performance).

I expect that wsgi's simple server mostly just forwards the request as-is to the wsgi app, so it'd be up to the app to implement all the http features needed.

**Comment by astronouth7303** [comment]
For the record, https://github.com/evansd/whitenoise/blob/master/whitenoise/responders.py is the most interesting bit of whitenoise.

**Comment by r0x0r** [comment]
@astronouth7303 this looks good to me.
If you manage to rip it out of whitenoise and make it work, while keeping it compact, I see no reason why not ditch Python's internal HTTP server.

**Comment by astronouth7303** [comment]
I'll get around to it!

**Comment by astronouth7303** [comment]
this stale bot is worse than the usual stale bot.

**Comment by astronouth7303** [comment]
could you please kindly reopen this?

**Comment by astronouth7303** [comment]
btw, @r0x0r how much do you want to discuss API design? What are your API stability requirements?

**Comment by r0x0r** [comment]
@astronouth7303 I prefer to avoid API rewrites, so best would be to get the API right in the first round. Misjudgements happen of course. 
I like the WSGI app approach, as long as the old way works as well, ie `create_window('title', url='index.html')`

**Comment by astronouth7303** [comment]
Yeah, I just ran through the examples and they all still work. (Or they're broken because of code I haven't touched.)

The way I've got the PR right now is that you can pass in a variety of things as your "url" (I added support for `pathlib.Path` while I was at it). If it gets something that looks like a wsgi app, it'll spin it up as a server.

The only downside is that the entrypoint to the app must be `/`.

**Comment by astronouth7303** [comment]
Is there a feature I missed?
